src/ApplicationInsights.Kubernetes/Extensions/ApplicationInsightsExtensions.cs

```diff
@@ -62,45 +58,59 @@ static ApplicationInsightsExtensions()
+            ILogger<KubernetesServiceCollectionBuilder> logger = null,
+            TimeSpan? timeout = null)
         {
+            logger = logger ?? serviceCollection.GetLogger<KubernetesServiceCollectionBuilder>();
+
             // 2 minutes by default maximum to wait for spinning up the container.
             timeout = timeout ?? TimeSpan.FromMinutes(2);
 
-            // According to the code, adding logging will not overwrite existing logging classes
-            // https://github.com/aspnet/Logging/blob/c821494678a30c323174bea8056f43b93a3ca6f4/src/Microsoft.Extensions.Logging/LoggingServiceCollectionExtensions.cs
-            // Becuase it uses 'TryAdd()' extenion method on service collection.
-            serviceCollection.AddLogging();
-
-            _logger.LogInformation(Invariant($"ApplicationInsights.Kubernetes.Version:{SDKVersionUtils.Instance.CurrentSDKVersion}"));
+            logger.LogInformation(Invariant($"ApplicationInsights.Kubernetes.Version:{SDKVersionUtils.Instance.CurrentSDKVersion}"));
             try
             {
-                serviceCollection = BuildK8sServiceCollection(serviceCollection, timeout.Value, kubernetesServiceCollectionBuilder);
+                serviceCollection = BuildK8sServiceCollection(serviceCollection, timeout.Value, k8sDetector, logger, kubernetesServiceCollectionBuilder);
             }
             catch (Exception ex)
             {
-                _logger.LogError("Failed to fetch ApplicaitonInsights.Kubernetes' info. Details " + ex.ToString());
+                logger.LogError("Failed to fetch ApplicaitonInsights.Kubernetes' info. Details " + ex.ToString());
             }
             return serviceCollection;
         }
 
         private static IServiceCollection BuildK8sServiceCollection(
             IServiceCollection services,
             TimeSpan timeout,
+            K8sDetector k8sDetector,
+            ILogger<KubernetesServiceCollectionBuilder> logger,
             IKubernetesServiceCollectionBuilder kubernetesServiceCollectionBuilder = null)
         {
-            kubernetesServiceCollectionBuilder = kubernetesServiceCollectionBuilder ?? new KubernetesServiceCollectionBuilder();
+            k8sDetector = k8sDetector ?? K8sDetectorFactory.Instance.CreateDetector();
+            kubernetesServiceCollectionBuilder = kubernetesServiceCollectionBuilder ?? new KubernetesServiceCollectionBuilder(k8sDetector, logger);
             services = kubernetesServiceCollectionBuilder.InjectServices(services, timeout);
             return services;
         }
+
+        /// <summary>
+        /// Gets a logger for given type.
+        /// Note: This method leads to build service provider during the injection of services and shall be avoid whenever possible.
+        /// </summary>
+        private static ILogger<T> GetLogger<T>(this IServiceCollection services)
+        {
+            // According to the code, adding logging will not overwrite existing logging classes
```

Comment:
Just say that "`AddLogging` method is idempotent, safe to call multiple times on the same collection"

Reply:
Sure. Firstly, learn a new word today - idempotent; secondly, the word itself seems a bit confused. Third, I'll simplify the comment in the next iteration.

Reply:
https://stackoverflow.com/questions/1077412/what-is-an-idempotent-operation